[SER] Patch 1: HitObject type lowering and SM 6.9 enablement

simoll opened the pull request:
Reduction of the complete SER implementation to just the HitObject type and its default constructor.

This has most of the infrastructure changes in DXC to support SER, eg static member functions for builtins, HitObject scalar type.

Specification PR: https://github.com/microsoft/hlsl-specs/pull/277

simoll commented:
@microsoft-github-policy-service agree company="NVIDIA"

bob80905 commented on tools/clang/test/SemaHLSL/hlsl/objects/HitObject/hitobject-entry-errors.hlsl:
```diff
@@ -0,0 +1,57 @@
+struct Attribs { float2 barys; };
+
+void UseHitObject() {
+// expected-error@+4{{Shader kind 'compute' incompatible with shader execution reordering (has to be raygeneration, closesthit or miss)}}
```
Does this test work if you move these expected errors into the corresponding entry functions? If not, would you re-order the entry functions below so that you can move these into the entry functions? Just a design preference, so that it's obvious the error is that it's the call site being in the context of a bad shader kind, not the function definition itself.

simoll replied:
Moved into entry functions in [57c861a](https://github.com/microsoft/DirectXShaderCompiler/pull/7097/commits/57c861aab39debbbe5ba53dbb2a9bb4a7c40a822).

bob80905 commented on tools/clang/test/SemaHLSL/hlsl/objects/HitObject/hitobject-in-buffer.hlsl:
```diff
@@ -0,0 +1,8 @@
+// expected-error@+1{{'HitObject' is an object and cannot be used as a type parameter}}
+RWStructuredBuffer<HitObject> InvalidBuffer;
+
+[shader("raygeneration")] void main() {
```
Can this function be removed?

simoll replied:
Removed in [57c861a](https://github.com/microsoft/DirectXShaderCompiler/pull/7097/commits/57c861aab39debbbe5ba53dbb2a9bb4a7c40a822)

simoll commented on tools/clang/lib/AST/HlslTypes.cpp:
```diff
@@ -613,6 +613,22 @@ bool IsHLSLResourceType(clang::QualType type) {
   return false;
 }
 
+bool IsHLSLHitObjectType(QualType type) {
```
The patch uses the same naming and implementation approach for `HitObject` as was done  for `NodeOutput` and `RayQuery`. The name `IsHLSLHitObjectType` says what the function does not how it is implemented, so i don't find that argument very convincing.
Do we really want to change this?

llvm-beanz replied:
We should not string match names. We have done this in the past, it was an unfortunate decision.

With the NodeRecord types we demonstrated a new solution, where we add an unspellable attribute to the type that identifies it as a specific object. We should do something like that here as well.

If you take a look at how `GetNodeIOType` is implemented that will give you a model for how this should be done.

simoll replied:
Fyi @bob80905 , commit bf3400a99 changes this function to rely on the `HLSLHitObjectAttr` instead.

simoll replied:
With  b028aaf6b this is using DXHitObjectAttr and no string matching. Marking as resolved.

simoll commented on tools/clang/lib/CodeGen/CGHLSLMSFinishCodeGen.cpp:
```diff
@@ -2794,6 +2794,56 @@ unsigned AlignBufferOffsetInLegacy(unsigned offset, unsigned size,
+             "Wrong signature for apparent HitObject constructor");
+
+    DXASSERT(!HitObjectCtor,
+             "Multiple candidate functions that quality as HitObject "
```
Fixed in [57c861a](https://github.com/microsoft/DirectXShaderCompiler/pull/7097/commits/57c861aab39debbbe5ba53dbb2a9bb4a7c40a822)

llvm-beanz commented on tools/clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -3967,6 +3999,13 @@ class HLSLExternalSource : public ExternalSemaSource {
     return IsRayQueryBasicKind(GetTypeElementKind(type));
   }
 
+  bool IsHitObjectBasicKind(ArBasicKind kind) {
```
This function seems unhelpful, so I don't really think we should have it.

simoll replied:
Removed in bf3400a99

llvm-beanz commented on tools/clang/lib/CodeGen/CGHLSLMSFinishCodeGen.cpp:
```diff
@@ -2794,6 +2794,56 @@ unsigned AlignBufferOffsetInLegacy(unsigned offset, unsigned size,
 }
 
+// Translate HitObject constructor to dx.op.hitObject_MakeNop
+void TranslateHitObjectConstructor(HLModule &HLM) {
```
This concerns me. CGHLSLMS is a giant hack that is nothing but a problem.

Is there a reason we can't have the AST represent a sane constructor that we can codegen directly?

simoll replied:
> With the NodeRecord types we demonstrated a new solution, where we add an unspellable attribute to the type that identifies it as a specific object. We should do something like that here as well.

bf3400a99 follows your suggestion and adds the `HLSLHitObjectAttr` attribute to model the HitObject type analogously to how it's done with `GetNodeIOType`

> This concerns me. CGHLSLMS is a giant hack that is nothing but a problem.
> Is there a reason we can't have the AST represent a sane constructor that we can codegen directly?

Luckily, `EmitHLSLScalarObjectDefaultConstructor` directly emits a call to `HitObject_MakeNop`. This change in CGHLSLMS was never needed and i removed it in 29c8705e5.

simoll replied:
This is obsolete, 358942da6 removes the special handling for the HitObject type in CGHLSLMS and elsewhere.

llvm-beanz commented on tools/clang/lib/SPIRV/AstTypeProbe.cpp:
```diff
@@ -1101,6 +1101,13 @@ bool isOpaqueType(QualType type) {
+
+    if (name == "HitObject") {
+      const CXXRecordDecl *typeRecordDecl = type->getAsCXXRecordDecl();
+      return typeRecordDecl &&
+             typeRecordDecl
+                 ->isImplicit(); // This is only our builtin, if SER is enabled
```
```suggestion
      // This is only our builtin, if SER is enabled
      return typeRecordDecl && typeRecordDecl->isImplicit(); 
```

simoll replied:
Accepted formatting suggestion in 0aa26b00a

llvm-beanz commented on tools/clang/lib/AST/ASTContextHLSL.cpp:
```diff
@@ -1161,6 +1163,31 @@ CXXRecordDecl *hlsl::DeclareRayQueryType(ASTContext &context) {
 
+CXXRecordDecl *hlsl::DeclareHitObjectType(ASTContext &Context) {
+  // HitObject { ... }
+  BuiltinTypeDeclBuilder TypeDeclBuilder(Context.getTranslationUnitDecl(),
```
Since we are not providing a Vulkan implementation of this, we should ensure that all the SER API additions are added to the `dx` namespace.

simoll replied:
This is in line with all other type declarations living in the hlsl namespace (and there is no dx namespace). The spec makes HitObject part of HLSL, not of the DX backend. Please elaborate what you mean here.

llvm-beanz replied:
Thank you for pointing out the issue with the spec. I've filed this issue tracking updating the spec: https://github.com/microsoft/hlsl-specs/issues/395

The original feedback stands.

simoll replied:
> With the NodeRecord types we demonstrated a new solution, where we add an unspellable attribute to the type that identifies it as a specific object. We should do something like that here as well.
> 
> If you take a look at how GetNodeIOType is implemented that will give you a model for how this should be done.

HitObject is now modelled after the NodeRecord types as you suggested.

llvm-beanz replied:
> [b028aaf](https://github.com/microsoft/DirectXShaderCompiler/commit/b028aaf6b849d285f307dd7077e68e8bc46a4a56) moves the SER implementation to the `dx` namespace. @llvm-beanz

This change does the wrong thing. You've moved the implementation to a namespace which is not what was requested. The `HitObject` type should be emitted into the AST under the `dx` namespace.

simoll replied:
[cefafa6](https://github.com/microsoft/DirectXShaderCompiler/pull/7097/commits/cefafa6a7d78fcff5488bac60cc848266c98e265)  moves the SER HitObject type and builtins to the `dx` namespace

llvm-beanz commented on tools/clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -1902,7 +1904,7 @@ void CodeGenFunction::EmitFunctionProlog(const CGFunctionInfo &FI,
     case ABIArgInfo::Direct: {
       // HLSL Change Begins
-      if (hlsl::IsHLSLMatType(Ty)) {
+      if (hlsl::IsHLSLMatType(Ty) || hlsl::IsHLSLHitObjectType(Ty)) {
```
This special handling seems odd. Why is this needed?

simoll replied:
This is necessary because otherwise call lowering will flatten the HitObject struct.

llvm-beanz replied:
See my other comment, but we should not be matching the pattern set by matrices because it is an endless source of bugs.

simoll replied:
Removed in 358942da6 . HitObject is now an aggregate.

llvm-beanz commented on tools/clang/lib/CodeGen/CGExprScalar.cpp:
```diff
@@ -1101,6 +1103,17 @@ void ScalarExprEmitter::EmitBinOpCheck(
 //                            Visitor Methods
 //===----------------------------------------------------------------------===//
 
+Value *ScalarExprEmitter::VisitCXXConstructExpr(CXXConstructExpr *E) {
```
This is very wrong. Scalar objects shouldn't have constructors.

simoll replied:
`HitObject` is modelled as a `TEK_Scalar` for codegen as it's done for the HLSL matrix type. This means `HitObject` flows as an opaque value and we only get trivial load/store/alloca of `HitObject` directly from codegen, no type replacement etc needed, exactly as we want it.
This is why you are seeing this constructor call visited by the `ScalarExprEmitter`.


Details on HitObject initialization:
- `HitObject` is declared with a default constructor. 
- When Clang builds an `InitializationSequence` for `HitObject`-typed things, `InitializeInitSequenceForHLSL(..)` adds the constructor call for an `SK_ConstructorInitialization` step.
- Since `HitObject` is `TEK_Scalar`, `ScalarExprEmitter` visits the constructor calls for `HitObject`. It directly emits `HitObject_MakeNop` through `EmitHLSLScalarObjectDefaultConstructor`.

llvm-beanz replied:
This and the handling in CGCall are wrong.

The way matrices are implemented in DXC are a gigantic hack that has caused a never-ending stream of bugs. We should not be following that pattern.

Instead we `HitObjects` should be treated like resource types, which are types that contain handles that we can safely treat as aggregates.

simoll replied:
358942da6 changes HitObject to be handled like a resource type. I've removed the scalar special handling cases.

damyanp commented on tools/clang/lib/Sema/SemaHLSLDiagnoseTU.cpp:
```diff
@@ -537,6 +562,19 @@ void hlsl::DiagnoseTranslationUnit(clang::Sema *self) {
     }
+
+    // Shader Execution Reordering
+    // ReorderThread(..) handled elsewhere.
```
```suggestion
    // MaybeReorderThread(..) handled elsewhere.
```

simoll replied:
Renamed to MaybeReorderThread in [36c1470](https://github.com/microsoft/DirectXShaderCompiler/pull/7097/commits/36c1470742bbb7677eed566e39e96aa1b4a54e00).  Resolved.

damyanp commented on tools/clang/test/SemaHLSL/hlsl/intrinsics/ser/reorder-unclaimed-pre-sm69.hlsl:
```diff
@@ -0,0 +1,13 @@
+// Check that intrinsic names of Shader Execution Reordering are unclaimed pre SM 6.9.
+// expected-no-diagnostics
+
+void ReorderThread(uint CoherenceHint, uint NumCoherenceHintBitsFromLSB) {
```
```suggestion
void MaybeReorderThread(uint CoherenceHint, uint NumCoherenceHintBitsFromLSB) {
```

simoll replied:
Renamed to MaybeReorderThread in [36c1470](https://github.com/microsoft/DirectXShaderCompiler/pull/7097/commits/36c1470742bbb7677eed566e39e96aa1b4a54e00).  Resolved.

damyanp commented on tools/clang/test/SemaHLSL/hlsl/intrinsics/ser/reorder-unclaimed-pre-sm69.hlsl:
```diff
@@ -0,0 +1,13 @@
+
+[shader("raygeneration")]
+void main() {
+  ReorderThread(15u, 4u);
```
```suggestion
  MaybeReorderThread(15u, 4u);
```

damyanp reviewed (changes requested):
Note, as per https://github.com/microsoft/hlsl-specs/pull/378, `ReorderThread` needs to be named `MaybeReorderThread`.

llvm-beanz commented on .cache/clangd/index/dxcmain.cpp.6E359381028004C7.idx:
Please remove the clangd indices.

simoll replied:
Removed

llvm-beanz commented on tools/clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -7966,6 +7966,16 @@ def err_hlsl_unsupported_semantic_index: Error<
+def err_hlsl_ser_invalid_shader_kind : Error<
+   "Shader kind '%0' incompatible with shader execution reordering (has to be raygeneration, closesthit or miss)">;
+def err_hlsl_reorder_invalid_shader_kind : Error<
+   "Shader kind '%0' incompatible with MaybeReorderThread intrinsic (only in raygeneration)">;
```
We should conditionalize this error to reduce how many new warnings we add. Changing the warning text to something like this would allow using a single diagnostic for diagnosing disallowed uses of `HitObject` and `MaybeReorderThread`.

```suggestion
   "%0 is unavailable in shader stage '%1 (requires %2)">;
```

simoll replied:
Applied the suggestion in [ba11d4f](https://github.com/microsoft/DirectXShaderCompiler/pull/7097/commits/ba11d4f3ffde25c5945852099caf9e983b2deb58)

llvm-beanz commented on tools/clang/lib/CodeGen/CodeGenTypes.cpp:
```diff
@@ -365,8 +367,9 @@ llvm::Type *CodeGenTypes::ConvertType(QualType T) {
                  .getTypePtr();
-      }
-      else
+      } else if (hlsl::IsHLSLHitObjectType(T)) {
+        return hlsl::dxilutil::GetHLSLHitObjectType(&TheModule);
+      } else
```
https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements
```suggestion
      } else if (hlsl::IsHLSLHitObjectType(T))
        return hlsl::dxilutil::GetHLSLHitObjectType(&TheModule);
      else
```

simoll replied:
Applied in [ba11d4f](https://github.com/microsoft/DirectXShaderCompiler/pull/7097/commits/ba11d4f3ffde25c5945852099caf9e983b2deb58)

llvm-beanz commented on tools/clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -776,11 +776,18 @@ Sema::ClassifyName(Scope *S, CXXScopeSpec &SS, IdentifierInfo *&Name,
+  // actual intrinsic declaration in AddOverloadedCallCandidates.
+  NamespaceDecl *nsDecl =
+      SS.isValid() ? SS.getScopeRep()->getAsNamespace() : nullptr;
+  bool isDxNamespace = nsDecl && nsDecl->getIdentifier() &&
```
This code change is concerning. Needing custom name classification rules for the `dx` namespace seems wrong.

What declaration are you hitting that got you here?

simoll replied:
Removed in [2bc823b](https://github.com/microsoft/DirectXShaderCompiler/pull/7097/commits/2bc823b0df962d343ea43e8be1a6246c67474f70). 'dx' builtins now implemented the same as 'vk' ones.

llvm-beanz commented on tools/clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -4809,11 +4996,17 @@ class HLSLExternalSource : public ExternalSemaSource {
+        NamespaceDecl *nsDecl = m_hlslNSDecl;
+        if (isVkNamespace)
+          nsDecl = m_vkNSDecl;
+        else if (isDxNamespace) {
+          nsDecl = m_dxNSDecl;
+        }
```
https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements
```suggestion
        else if (isDxNamespace)
          nsDecl = m_dxNSDecl;
```

simoll replied:
Applied in [ba11d4f](https://github.com/microsoft/DirectXShaderCompiler/pull/7097/commits/ba11d4f3ffde25c5945852099caf9e983b2deb58)

llvm-beanz commented on tools/clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -4757,10 +4922,17 @@ class HLSLExternalSource : public ExternalSemaSource {
+    // Exceptions:
+    // - Vulkan-specific intrinsics live in the 'vk::' namespace.
+    // - DirectX-specific intrinsics live in the 'dx::' namespace.
+    if (isQualified && !isGlobalNamespace && !isVkNamespace && !isDxNamespace) {
       return false;
     }
```
https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements
```suggestion
    if (isQualified && !isGlobalNamespace && !isVkNamespace && !isDxNamespace)
      return false;
```

simoll replied:
Applied in [ba11d4f](https://github.com/microsoft/DirectXShaderCompiler/pull/7097/commits/ba11d4f3ffde25c5945852099caf9e983b2deb58)

llvm-beanz commented on tools/clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -10936,7 +10936,13 @@ bool Sema::buildOverloadedCallSet(Scope *S, Expr *Fn,
+        ULE->getQualifier()->getAsNamespace()->getName() == "dx";
+
+    assert((!ULE->getQualifier() || isVkNamespace || isDxNamespace) &&
+           "non-vk qualified name with ADL");
```
```suggestion
           "non-namespace qualified name with ADL");
```

simoll replied:
Applied in [ba11d4f](https://github.com/microsoft/DirectXShaderCompiler/pull/7097/commits/ba11d4f3ffde25c5945852099caf9e983b2deb58)

simoll commented:
> Note, as per [microsoft/hlsl-specs#378](https://github.com/microsoft/hlsl-specs/pull/378), `ReorderThread` needs to be named `MaybeReorderThread`.

Right, `ReorderThread` is now renamed to `MaybeReorderThread` everywhere

llvm-beanz commented on tools/clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -7970,6 +7970,14 @@ def err_hlsl_unsupported_semantic_index: Error<
   "'%0' is defined with semantic index %1, but only values 0 through %2 are supported">;
 // HLSL Change Ends
 
+// SER Change Starts
+// Shader Execution Reordering
```
```suggestion
// HLSL Change Starts - Shader Execution Reordering
```

Alternatively you could remove the HLSL Change Ends comment above.

simoll replied:
Removed the HLSL Change end marker in latest version

llvm-beanz commented on tools/clang/lib/CodeGen/CodeGenTypes.cpp:
```diff
@@ -14,21 +14,23 @@
 #include "CGRecordLayout.h"
+#include "CodeGenModule.h" // HLSL Change
 #include "TargetInfo.h"
+#include "dxc/DXIL/DxilUtil.h"
```
```suggestion
#include "dxc/DXIL/DxilUtil.h" // HLSL Change
```

llvm-beanz commented on tools/clang/lib/CodeGen/CodeGenTypes.cpp:
```diff
@@ -14,21 +14,23 @@
 #include "clang/AST/DeclCXX.h"
-#include "clang/AST/DeclTemplate.h"
 #include "clang/AST/DeclObjC.h"
+#include "clang/AST/DeclTemplate.h"
```
I know this isn't really a change so much as clang-format, but marking it makes comparing to Clang 3.7 easier.

```suggestion
#include "clang/AST/DeclTemplate.h" // HLSL Change - clang-format
```

llvm-beanz commented on tools/clang/lib/CodeGen/CodeGenTypes.cpp:
```diff
@@ -14,21 +14,23 @@
 #include "clang/AST/DeclObjC.h"
+#include "clang/AST/DeclTemplate.h"
 #include "clang/AST/Expr.h"
+#include "clang/AST/HlslTypes.h"
```
```suggestion
#include "clang/AST/HlslTypes.h" // HLSL Change
```

llvm-beanz commented on tools/clang/lib/CodeGen/CodeGenTypes.cpp:
```diff
@@ -365,7 +367,8 @@ llvm::Type *CodeGenTypes::ConvertType(QualType T) {
                                        ArrayType::ArraySizeModifier::Normal, 0)
                  .getTypePtr();
-      }
+      } else if (hlsl::IsHLSLHitObjectType(T))
+        return hlsl::dxilutil::GetHLSLHitObjectType(&TheModule);
```
```suggestion
      } else if (hlsl::IsHLSLHitObjectType(T)) // HLSL Change
        return hlsl::dxilutil::GetHLSLHitObjectType(&TheModule);
```

llvm-beanz commented on tools/clang/lib/CodeGen/TargetInfo.cpp:
```diff
@@ -17,6 +17,7 @@
 #include "CGCXXABI.h"
 #include "CGValue.h"
 #include "CodeGenFunction.h"
+#include "clang/AST/HlslTypes.h"
```
```suggestion
```

llvm-beanz commented on tools/clang/lib/Sema/SemaCXXScopeSpec.cpp:
```diff
@@ -11,15 +11,16 @@
 #include "clang/AST/ASTContext.h"
 #include "clang/AST/DeclTemplate.h"
 #include "clang/AST/ExprCXX.h"
+#include "clang/AST/HlslTypes.h"
```
```suggestion
#include "clang/AST/HlslTypes.h" // HLSL Change
```

llvm-beanz commented on tools/clang/lib/Sema/SemaCXXScopeSpec.cpp:
```diff
@@ -11,15 +11,16 @@
 #include "clang/Basic/PartialDiagnostic.h"
 #include "clang/Sema/DeclSpec.h"
 #include "clang/Sema/Lookup.h"
+#include "clang/Sema/SemaInternal.h"
```
```suggestion
#include "clang/Sema/SemaInternal.h" // HLSL Change - clang-format
```

llvm-beanz commented on tools/clang/lib/Sema/SemaCXXScopeSpec.cpp:
```diff
@@ -200,9 +201,13 @@ bool Sema::RequireCompleteDeclContext(CXXScopeSpec &SS,
-  if (tagType && tagType->isBeingDefined())
+  if (tagType && tagType->isBeingDefined() &&
+      (hlsl::IsUserDefinedRecordType(type) ||
+       hlsl::IsHLSLObjectWithImplicitMemberAccess(type)))
```
This change doesn't make sense to me. Can you explain why this is required?

If we get here and `tagType` a builtin type that is not yet completed, we should go down into `RequireCompleteType` below which will call into the external sema source to complete it. I'm confused why we need additional early-outs here, and the comment above makes it more confusing.

Also changes in this file should be wrapped in `// HLSL Change Begin|End` comments.

simoll replied:
The current HLSL-specific modifications to scoped lookup do not support 'static'. I've changed my approach and eagerly complete the HitObject type (including its static functions) instead. This means, no special handling for HitObject is any longer necessary.
This specific change is gone now - considering this resolved.

llvm-beanz commented on tools/clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -804,7 +804,7 @@ Sema::ClassifyName(Scope *S, CXXScopeSpec &SS, IdentifierInfo *&Name,
         return BuildDeclarationNameExpr(SS, Result, /*ADL=*/false);
       }
     }
-    
```
Please revert this whitespace only change.

llvm-beanz commented on tools/clang/test/SemaHLSL/hlsl/objects/HitObject/hitobject-target-profile-error.hlsl:
```diff
@@ -0,0 +1,9 @@
+// RUN: not %dxc -T lib_6_8 %s 2>&1 | FileCheck %s
+
+namespace dx {}
```
What happens if you remove this? That would be the more common situation.

simoll replied:
Kept this test - but added another one that specifically checks that a namespace 'dx' is not predeclared pre-SM6.9

simoll commented:
> @simoll - do you know why all the builds are failing for this PR?

It's because i've rebased the PR onto tot and tot was failing at that time:
`  FAIL: Clang :: CodeGenDXIL/hlsl/intrinsics/dot.hlsl (4 of 2488)`

simoll commented on tools/clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -11529,6 +11746,60 @@ static bool isStringLiteral(QualType type) {
+  }
+
+  // Return for anything that is not a HitObject intrinsic.
+  std::string FeatureStr = "Shader Execution Reordering";
```
Using %select in latest version

simoll commented on lib/HLSL/HLOperationLower.cpp:
```diff
@@ -6064,6 +6064,24 @@ Value *TranslateUnpack(CallInst *CI, IntrinsicOp IOP, OP::OpCode opcode,
+                              HLOperationLowerHelper &helper,
+                              HLObjectOperationLowerHelper *pObjHelper,
+                              bool &Translated) {
+  return UndefValue::get(CI->getType()); // TODO: Merge SER DXIL patches
```
Added test stubs in latest revision for MaybeReorderThread, MakeNop and the HitObject type

simoll commented:
> Missing tests:

I've added the requestedtests:

> 
> 1. IR test for the HLOperationLowering changes
- tools\clang\test\HLSLFileCheck\hlsl\objects\HitObject\maybereorder.hlsl
- tools\clang\test\HLSLFileCheck\hlsl\objects\HitObject\hitobject_make.hlsl

> 2. HitObject and new APIs used without `dx` namespace (should error)
tools\clang\test\SemaHLSL\hlsl\objects\HitObject\maybereorderthread-without-namespace.hlsl
- tools\clang\test\SemaHLSL\hlsl\objects\HitObject\hitobject-without-namespace.hlsl

> 3. HitObject and new APIs used with `using namespace dx` instead of explicit namespace
- tools\clang\test\SemaHLSL\hlsl\objects\HitObject\hitobject-using-namespace.hlsl

llvm-beanz commented:
> > Missing tests:
> 
> I've added the requested tests:
> 
> > 1. IR test for the HLOperationLowering changes
> 
> * tools\clang\test\HLSLFileCheck\hlsl\objects\HitObject\maybereorder.hlsl
> * tools\clang\test\HLSLFileCheck\hlsl\objects\HitObject\hitobject_make.hlsl

These are HLSL tests, we should have IR tests for IR passes. IR tests start as IR and verify transformations in isolation.

Please see the tests here as examples: https://github.com/microsoft/DirectXShaderCompiler/tree/main/tools/clang/test/DXC/Passes

llvm-beanz commented:
@simoll the bots are failing on one of your tests:
> D:\a\1\s\tools\clang\test\HLSL\..\HLSLFileCheck\hlsl\objects\HitObject\maybereorder.hlsl [Failed]

simoll commented:
> 
> These are HLSL tests, we should have IR tests for IR passes. IR tests start as IR and verify transformations in isolation.
> 
> Please see the tests here as examples: https://github.com/microsoft/DirectXShaderCompiler/tree/main/tools/clang/test/DXC/Passes

Added dxilgen pass tests that cover all HitObject types and intrinsics in this PR:
* tools\clang\test\DXC\Passes\DxilGen\maybereorder_dxilgen.ll
* tools\clang\test\DXC\Passes\DxilGen\hitobject_dxilgen.ll

> @simoll the bots are failing on one of your tests:

Fixed in latest version. Output mismatch between release and debug build leading to CHECK: clauses not matching in release

@llvm-beanz - back to you

llvm-beanz reviewed (approved):
I'm fine with this as it is now. We'll need to shore up the IR testing as subsequent PRs come in.

@tex3d or @pow2clk, any feedback?

damyanp commented:
@simoll - while we're waiting on @tex3d or @pow2clk's review, I notice that there's a merge conflict that needs resolving.

tex3d reviewed (changes requested):
Partial review, but there are a couple blocking issues.
- A breaking change to HLSL_INTRINSICS should be separated and made by MSFT to synchronize internal dependencies, see #7198
- I have requested changing to availability attributes for shader model version check which will require additional changes described by issue #7198 and incoming changes to check the availability attributes from #7156. This should eliminate the need for the special `EnableShaderExecutionReordering` flag added to various option classes as well.
- I'm curious why this has broken with the pattern for object CodeGen and later lowering to DXIL types used with other object types.

simoll commented on utils/hct/hctdb_instrhelp.py:
```diff
@@ -971,7 +972,7 @@ def get_hlsl_intrinsics():
             last_ns = i.ns
             id_prefix = (
-                "IOP" if last_ns == "Intrinsics" or last_ns == "VkIntrinsics" else "MOP"
+                "IOP" if last_ns.endswith("Intrinsics") else "MOP"
```
Removed in df1e26fd4

simoll commented on tools/clang/include/clang/Basic/LangOptions.h:
```diff
@@ -162,6 +162,7 @@ class LangOptions : public LangOptionsBase {
   bool EnableDX9CompatMode = false;
   bool EnableFXCCompatMode = false;
   bool EnablePayloadAccessQualifiers = false;
+  bool EnableShaderExecutionReordering = false;
```
Removed SER opts in c390f311e

simoll commented on tools/clang/include/clang/Frontend/CodeGenOptions.h:
```diff
@@ -244,6 +244,8 @@ class CodeGenOptions : public CodeGenOptionsBase {
   /// Put shader sources and options in the module
   bool HLSLEmbedSourcesInModule = false;
+  /// Enable Shader Execution Reordering.
+  bool HLSLEnableShaderExecutionReordering = false;
```
Using SM instead of opts in c390f311e

simoll commented on tools/clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -3591,6 +3749,12 @@ class HLSLExternalSource : public ExternalSemaSource {
         recordDecl = DeclareRayQueryType(*m_context);
+      } else if (kind == AR_OBJECT_HIT_OBJECT) {
+        // Declare 'HitObject' in '::dx' extension namespace.
+        if (EnableSER) {
```
Blocked until https://github.com/microsoft/DirectXShaderCompiler/pull/7156 has panded

simoll replied:
Same as for the other instance: always declaring the type and using availability attributes instead.

simoll commented on tools/clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -3894,6 +4071,9 @@ class HLSLExternalSource : public ExternalSemaSource {
       AddIntrinsicTableMethods(intrinsic);
     }
 
+    if (SM->IsSM69Plus())
```
Blocked until https://github.com/microsoft/DirectXShaderCompiler/pull/7156 has landed

simoll replied:
Adding the intrinsic unconditionally with an AvailabilityAttribute now. Considering this resolved.

damyanp commented:
@simoll - just to let you know, #7156 was resolved via #7199.  I think #7156 also has some parts that might be of interest to you wrt availability attributes.  @tex3d can probably fill in more details.

simoll commented:
> @simoll - just to let you know, #7156 has landed. #7199 also has some parts that might be of interest to you wrt availability attributes. @tex3d can probably fill in more details.

Great! I'll rebase and make changes to use the availability attributes

simoll commented on include/dxc/dxcapi.internal.h:
```diff
@@ -165,6 +167,7 @@ struct HLSL_INTRINSIC {
   BOOL bReadOnly;          // Only read memory
   BOOL bReadNone;          // Not read memory
   BOOL bIsWave;            // Is a wave-sensitive op
+  BOOL bStaticMember;      // HLSL static member function
```
I made the following changes after the PR was merged:
- introduced a new `INTRIN_FLAG_STATIC_MEMBER` to represent static  member functions
- added min_sm=6.9 attribute to the SER intrinsic declarations

simoll commented on tools/clang/lib/CodeGen/CodeGenTypes.cpp:
```diff
@@ -365,7 +367,8 @@ llvm::Type *CodeGenTypes::ConvertType(QualType T) {
                  .getTypePtr();
-      }
+      } else if (hlsl::IsHLSLHitObjectType(T))
+        return hlsl::dxilutil::GetHLSLHitObjectType(&TheModule);
```
Is the suggestion to use the standard clang type codegen here and perform type replacement in the hlsl sroa pass?
If we remove this change, we assert in the ScalarReplAggregatesHLSL.cpp :: RewriteCall special case for RayQuery rewriting.

damyanp replied:
@simoll - can you file an issue on this and we can address this later.  Thanks!

simoll replied:
Yep, tracking this in issue #7233 - [SER] Do not short-cut to %dx.types.HitObject DXIL type in HL from AST codegen

simoll commented on include/dxc/DXIL/DxilUtil.h:
```diff
@@ -162,6 +162,8 @@ GetHLSLResourceProperties(llvm::Type *Ty);
 bool IsHLSLObjectType(llvm::Type *Ty);
 bool IsHLSLRayQueryType(llvm::Type *Ty);
+llvm::Type *GetHLSLHitObjectType(llvm::Module *M);
+bool IsHLSLHitObjectType(llvm::Type *Ty);
```
This is used by IsHLSLObjectType . I commented below on HitObject type lowering

bob80905 commented on tools/clang/lib/AST/ASTContextHLSL.cpp:
```diff
@@ -1243,6 +1243,10 @@ CXXRecordDecl *hlsl::DeclareHitObjectType(NamespaceDecl &NSDecl) {
 
+  // Shader Execution Reordering requires SM6.9
+  VersionTuple VT69 = VersionTuple(6, 9);
+  pConstructorDecl->addAttr(ConstructAvailabilityAttribute(Context, VT69));
```
Would you be able to add an ast-dump test to verify this new attribute in the AST? Same with the attribute you add in SemaHLSL.cpp now.
Separately, I'm not sure how this attribute is useful. Does `dx::HitObject::MakeNop();` create a constructor Decl? If not, I don't see where this attribute is actually generating an error when SM < 6.9, and I suspect the other attribute you added in SemaHLSL is actually doing the work.
For example, if you were to change this VT69 to, say, VT63, would any of the errors in your tests go away? Which ones would change?

simoll replied:
I've removed the AvailabilityAttribute from the constructor.. it was unused.

I've added ast tests for `HitObject::MakeNop` and `MaybeReorderThread` to check for the AvailabilityAttribute as you suggested.

bob80905 commented on tools/clang/test/SemaHLSL/hlsl/objects/HitObject/hitobject-unavailable-pre-sm69.hlsl:
```diff
@@ -1,20 +1,9 @@
 
 [shader("raygeneration")]
 void main() {
+  // expected-warning@+1{{potential misuse of built-in function 'dx::HitObject::MakeNop' in shader model lib_6_8; introduced in shader model 6.9}}
```
Not sure if HitObject::Miss is in yet or not, but I imagine a similar test would need to be added in that case. Just a comment FYI, no resolution needed.

simoll commented:
> > @simoll - just to let you know, #7156 has landed. #7199 also has some parts that might be of interest to you wrt availability attributes. @tex3d can probably fill in more details.
> 
> Great! I'll rebase and make changes to use the availability attributes

This is using AvailabilityAttributes now. There are ast tests to check for the presence of the attribute and there are tests to check for the availability diagnostic when SER is used pre-SM6.9.

damyanp commented on include/dxc/DXIL/DxilUtil.h:
```diff
@@ -162,6 +162,8 @@ GetHLSLResourceProperties(llvm::Type *Ty);
 bool IsHLSLResourceType(llvm::Type *Ty);
 bool IsHLSLObjectType(llvm::Type *Ty);
 bool IsHLSLRayQueryType(llvm::Type *Ty);
+llvm::Type *GetHLSLHitObjectType(llvm::Module *M);
```
This will be resolved as part of https://github.com/microsoft/DirectXShaderCompiler/pull/7097#discussion_r2003791199

damyanp commented on tools/clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -11554,6 +11845,8 @@ void Sema::DiagnoseReachableHLSLCall(CallExpr *CE, const hlsl::ShaderModel *SM,
   if (!IsBuiltinTable(IntrinsicAttr->getGroup()))
     return;
 
+  DiagnoseReachableCallForSER(CE, EntrySK, EntryDecl, SM);
```
@simoll  - we discussed this today, will make this change.

simoll replied:
Resolved, implemented as shown here

tex3d commented on tools/clang/lib/Sema/SemaHLSLDiagnoseTU.cpp:
```diff
@@ -430,13 +433,27 @@ class HLSLCallDiagnoseVisitor // Could rename to HLSLReachableDiagnoseVisitor
+    switch (ND->getKind()) {
+    case Decl::Function:
+    case Decl::FunctionTemplate:
+    case Decl::CXXMethod:
```
I think `isa<FunctionDecl>` is all we need.

tex3d commented on tools/clang/test/DXC/Passes/DxilGen/hitobject_dxilgen.ll:
```diff
@@ -0,0 +1,100 @@
+; RUN: %dxopt %s -hlsl-passes-resume -dxilgen -S | FileCheck %s
```
```suggestion
; RUN: %dxopt %s -hlsl-passes-resume -dxilgen -S | FileCheck %s
; REQUIRES: dxil-1-9
```

tex3d commented on tools/clang/test/DXC/Passes/DxilGen/maybereorder_dxilgen.ll:
```diff
@@ -0,0 +1,105 @@
+; RUN: %dxopt %s -hlsl-passes-resume -dxilgen -S | FileCheck %s
```
```suggestion
; RUN: %dxopt %s -hlsl-passes-resume -dxilgen -S | FileCheck %s
; REQUIRES: dxil-1-9
```

tex3d commented on tools/clang/test/HLSLFileCheck/hlsl/objects/HitObject/hitobject_make.hlsl:
```diff
@@ -0,0 +1,11 @@
+// RUN: %dxc -T lib_6_9 -E main %s | FileCheck %s
```
All these tests currently under `HLSLFileCheck` can be moved under the `CodeGenDXIL` path to make them ordinary lit shell tests.  This is something we are generally advising for new tests that aren't required to run under this path due to special tooling only available here still.

When doing so, you need to add `// REQUIRES: dxil-1-9`

```suggestion
// RUN: %dxc -T lib_6_9 -E main %s | FileCheck %s
// REQUIRES: dxil-1-9
```

tex3d commented on tools/clang/test/SemaHLSL/hlsl/objects/HitObject/hitobject-in-buffer.hlsl:
```diff
@@ -0,0 +1,4 @@
+// RUN: %dxc -T lib_6_9 %s -verify
+
+// expected-error@+1{{'dx::HitObject' is an object and cannot be used as a type parameter}}
+RWStructuredBuffer<dx::HitObject> InvalidBuffer;
```
Should probably also block it from being used as a global constant buffer variable.

simoll replied:
Yes, but i see no infrastructure / pre-existing diagnostic in DXC to do this. I suggest we open an issue for this to address later.

simoll replied:
@tex3d - Tracking this change request in #7234 . Can we consider this non-blocking for this PR?

tex3d replied:
Yes, sounds good.

simoll commented on tools/clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -7662,8 +7662,8 @@ def warn_hlsl_rayquery_flags_conflict : Warning<
   "operator cannot be used with built-in type %0">;
-def warn_hlsl_builtin_constant_unavailable: Warning<
-  "potential misuse of built-in constant %0 in shader model %1; introduced"
+def warn_hlsl_builtin_unavailable: Warning<
```
This uses `warn_hlsl_intrinsic_in_wrong_shader_model` now . Removed the ser-specific diagnostic.

simoll commented on tools/clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -1816,15 +1941,14 @@ static void AddHLSLIntrinsicAttr(FunctionDecl *FD, ASTContext &context,
+    FD->addAttr(AvailabilityAttr::CreateImplicit(
+        context, &context.Idents.get(""), clang::VersionTuple(Major, Minor),
+        clang::VersionTuple(), clang::VersionTuple(), false,
+        "HLSL Intrinsic availability limited by shader model."));
```
> This results in less code and complexity than the current approach, I think.

Simplified following the latest suggestion.

> It turns out this string will be unused, so we can leave it empty ("").

Replaced with "" and updated tests.

Resolved.

simoll commented on tools/clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -7991,6 +7991,12 @@ def warn_hlsl_legacy_integer_literal_signedness: Warning<
+
+// Shader Execution Reordering
+def err_hlsl_ser_unsupported : Error<
+   "%select{dx::HitObject|dx::MaybeReorderThread}0 is unavailable in shader stage '%1' (requires 'raygeneration'%select{|, 'closesthit' or 'miss'}2)">;
```
There are two diagnostics now as suggested.

simoll commented on tools/clang/lib/Sema/SemaHLSLDiagnoseTU.cpp:
```diff
@@ -430,13 +433,27 @@ class HLSLCallDiagnoseVisitor // Could rename to HLSLReachableDiagnoseVisitor
+      break;
     }
+
+    sema->Diag(Loc, diag::warn_hlsl_builtin_unavailable)
```
Using that diagnostic now and removed the SER-specific diagnostic.

simoll commented:
@tex3d  - you made the suggestion to move some SER tests out of HLSLFileCheck to CodeGenDXIL (the exact comment got lost AFAICT). Is this necessary?

tex3d commented:
> @tex3d - you made the suggestion to move some SER tests out of HLSLFileCheck to CodeGenDXIL (the exact comment got lost AFAICT). Is this necessary?

No, not absolutely necessary.  Generally, we are trying to avoid adding new tests under HLSLFileCheck unless necessary for testing infrastructure reasons (D3DReflect testing, for instance).  We could create a follow-up issue for bulk test reorg later if it helps the flow of other commits on top of this one.

One thing that will help with that move is the addition of the `REQUIRES: dxil-1-9`.  Under HLSLFileCheck, there is automatic detection of dxc compilation shader target arguments, which causes skips if, for instance, the output is tested against a DXIL validator that doesn't meet the requirement.  IR opt tests should not require this, `-fcgl` tests should not require this, but ones that  compile to a shader target with `dxc` do.

tex3d commented on tools/clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -7665,6 +7665,9 @@ def err_hlsl_unsupported_builtin_op: Error<
   " in shader model %2">, InGroup<HLSLAvailabilityConstant>;
+def warn_hlsl_builtin_type_unavailable: Warning<
+  "potential misuse of built-in type %0 in shader model %1; introduced"
+  " in shader model %2">, InGroup<HLSLAvailability>;
```
Should we add `DefaultError` to make this an error by default?

simoll replied:
> Should we add DefaultError to make this an error by default?

I don't think so given that the type may be used on dead code paths in HLSL and optimized out before it is caught by the DXIL validator. The DXIL validator change we track in #7234

simoll replied:
Turned this into a `DefaultError` - we do the same for the other diags of this kinds (except bui;ltin constants).Resolved,

tex3d reviewed (approved):
Looks good!
